src/ameba/rule/style/heredoc_escape.cr

```diff
@@ -0,0 +1,68 @@
+module Ameba::Rule::Style
+  # A rule that enforces using an escaped heredoc to escape interpolation or escape sequences instead of backslashes.
+  #
+  # For example, this is considered invalid:
+  #
+  # ```
+  # <<-DOC
+  #   \#{:interpolated}
+  #   \\n
+  #   DOC
+  # ```
+  #
+  # And should be written as:
+  #
+  # ```
+  # <<-'DOC'
+  #   #{:interpolated}
+  #   \n
+  #   DOC
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Style/HeredocEscape:
+  #   Enabled: true
+  # ```
+  class HeredocEscape < Base
+    include AST::Util
+
+    properties do
+      since_version "1.7.0"
+      description "Recommends using the escape heredoc start when escaping interpolation or other chars in a heredoc body"
+    end
+
+    MSG_ESCAPE_NEEDED     = "Use an escaped heredoc"
+    MSG_ESCAPE_NOT_NEEDED = "Unnecessary heredoc escape"
+
+    def test(source, node : Crystal::StringInterpolation)
+      return unless (code = node_source(node, source.lines)) && code.starts_with?("<<-")
+
+      # Heredocs without interpolation are always size 1
+      return unless node.expressions.size == 1
+
+      return unless expr = node.expressions.first?.try(&.as?(Crystal::StringLiteral))
+
+      body = code.split('\n')[1..-2].join("\n")
```

Comment:
What's that needed for?

Reply:
We need the raw source code of the heredoc itself to differentiate between `\n` spelled out and an actual newline. It's not possible to do so with just the information provided on the node.